flowbench/BenchZstd.cpp

```diff
@@ -0,0 +1,242 @@
+/*
+ * BenchZstd.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "benchmark/benchmark.h"
+#include "flow/IRandom.h"
+
+#include <cstdio>
+#include <memory>
+#include <sstream>
+#include <fstream>
+
+#ifdef ZSTD_LIB_SUPPORTED
+
+#define ZSTD_STATIC_LINKING_ONLY
+#include "zstd.h"
+
+// Benchmark zstd performance. To use it:
+//  # export BM_ZSTD_DATA=path/to/datafile
+//  # bin/flowbench --benchmark_filter=bench_zstd
+
+// Compress with raw ZSTD API
+static inline std::string compress(const std::string& data, int level) {
+	const char* srcBegin = data.data();
+	size_t destSize = ZSTD_compressBound(data.size());
+	std::unique_ptr<char[]> dest = std::make_unique<char[]>(destSize);
+	size_t bytes = ZSTD_compress(dest.get(), destSize, srcBegin, data.size(), level);
+	return std::string(dest.get(), bytes);
+}
+
+// Compress with raw ZSTD API ZSTD_compress2
+static inline std::string compress2(const std::string& data, ZSTD_CCtx* cctx) {
+	const char* srcBegin = data.data();
+	size_t destSize = ZSTD_compressBound(data.size());
+	std::unique_ptr<char[]> dest = std::make_unique<char[]>(destSize);
+	size_t bytes = ZSTD_compress2(cctx, dest.get(), destSize, srcBegin, data.size());
+	return std::string(dest.get(), bytes);
+}
+
+// Deompress with raw ZSTD API
+static inline std::string decompress(const std::string& data) {
+	size_t destSize = ZSTD_decompressBound(data.data(), data.size());
+	std::unique_ptr<char[]> dest = std::make_unique<char[]>(destSize);
+	size_t bytes = ZSTD_decompress(dest.get(), destSize, data.data(), data.size());
+	return std::string(dest.get(), bytes);
```

Comment:
This `std::string` constructor creates a copy of the bytes in `dest`, which will affect performance. Returning the `std::unique_ptr<char[]>` directly can avoid this copy.

Reply:
fixed.